src/python/pants/engine/internals/rule_visitor.py

```diff
@@ -28,6 +35,86 @@ def _get_starting_indent(source: str) -> int:
     return 0
 
 
+def _node_str(node: Any) -> str:
+    if isinstance(node, ast.Name):
+        return node.id
+    if isinstance(node, ast.Attribute):
+        return ".".join([_node_str(node.value), node.attr])
+    if isinstance(node, ast.Call):
+        return _node_str(node.func)
+    if sys.version_info[0:2] < (3, 8):
+        if isinstance(node, ast.Str):
+            return node.s
+    else:
+        if isinstance(node, ast.Constant):
+            return str(node.value)
+    return str(node)
+
+
+class _TypeStack:
+    def __init__(self, module_name: str) -> None:
+        self.root = sys.modules[module_name]
+        self._stack: list[dict[str, Any]] = []
+        self.push(self.root)
+
+    def __repr__(self) -> str:
+        from pprint import pformat
+
+        return f"TypeStack<\n{pformat(self._stack)}\n>"
+
+    def __getitem__(self, name: str) -> Any:
+        for ns in reversed(self._stack):
+            if name in ns:
+                return ns[name]
+        return self.root.__builtins__.get(name, None)
+
+    def __setitem__(self, name: str, value: Any) -> None:
+        self._stack[-1][name] = value
+
+    def push(self, frame: object) -> None:
+        ns = dict(frame if isinstance(frame, dict) else frame.__dict__)
+        self._stack.append(ns)
+
+    def pop(self) -> None:
+        assert len(self._stack) > 1
+        self._stack.pop()
+
+
+def _lookup_annotation(obj: Any, attr: str) -> Any:
+    """Get type assocated with a particular attribute on object. This can get hairy, especially on
+    Python <3.10.
+
+    https://docs.python.org/3/howto/annotations.html#accessing-the-annotations-dict-of-an-object-in-python-3-9-and-older
+    For this reason, we've copied the `inspect.get_annotations` method from CPython `main` branch.
+    """
+    if hasattr(obj, attr):
+        return getattr(obj, attr)
+    else:
+        try:
+            return get_annotations(obj, eval_str=True).get(attr)
+        except (NameError, TypeError):
+            return None
```

Comment:
At least for a while, it would seem like a reasonable restriction to require types to be present on callees, and to fail inside `get_annotations` if they're not. Because the rule graph errors you're likely to get if you feed it a series of `Any` types would be pretty spectacular.

Reply:
Not sure I follow. If this returns `None`, we'll catch it later before being used, if at all (in `_check_constraint_arg_type()`)

If I've missed something, a small example demonstrating the case could help :)

Reply:
Ah, ok. That sounds reasonable then probably.

I think that I was expecting that if the callee was missing a signature, we would fail during extraction of the annotations, rather than closer to the caller... i.e., that we would blame the callee for not being annotated, rather than the caller for using something un-annotated.

Reply:
Well, now the un-annotated thing could be anything.. so we can't really blame that until we try to use it from a rule code.

Reply:
If we track the origin of the type, we could blame the callee for being unannotated with the context of where it's being called from, I'll look into that.

Reply:
Fixed for function return types by simply looking up the source for the offending function to include in the error message which already states what type and where the issue is coming from. See test cases for more context.

Example from a test case (using block quote for line wrapping):

> /private/var/folders/y2/6t5bbk554wv5h6w_qlgpcpnm0000gr/T/pants-sandbox-w0DH7J/src/python/pants/engine/internals/rule_visitor_test.py:285: Could not resolve type for `request.bad_meth` in module pants.engine.internals.rule_visitor_test.
>
> Return type annotation required for `bad_meth` in /private/var/folders/y2/6t5bbk554wv5h6w_qlgpcpnm0000gr/T/pants-sandbox-w0DH7J/src/python/pants/engine/internals/rule_visitor_test.py:239